File: trust-tasks-dart-tsp/lib/src/consumer.dart
import 'dart:typed_data';

import 'package:affinidi_tsp/affinidi_tsp.dart';
import 'package:trust_tasks/trust_tasks.dart';

import 'envelope.dart';

/// What [TspConsumer.receive] made of one TSP message.
final class Received<R> {
  const Received({
    required this.outcome,
    required this.transport,
    required this.reply,
  });

  /// The §7.2 outcome, exactly as `consumeInbound` defines it.
  final ConsumeOutcome<R> outcome;

  /// Who sent it (authenticated) and who received it.
  final TspTransport transport;

  /// The document to send back, as JSON, or null when nothing is to be sent:
  /// a fire-and-forget success, a suppressed `identityMismatch` (§8.1), or a
  /// duplicate with no retained response (§7.1). Seal it with
  /// [TspConsumer.packReply].
  final Map<String, dynamic>? reply;
}

/// The guarded inbound path for the TSP binding: resolve the sender, open the
/// sealed message, then run SPEC §7.2 with the duplicate-execution record of
/// item 11 and the freshness bound it needs — both **on by default**.
///
/// §7 records that TSP data messages "do not inherently prevent replay", and
/// under §5's routed and nested carriage an intermediary may re-send the sealed
/// inner message. So an ordinary re-forward — no attacker involved — executes a
/// consequential task twice unless the consumer keeps the record. The record is
/// keyed on the document `id`, never the TSP envelope, which is resealed with
/// fresh material on every send (§7.1).
///
/// **One consumer per process, not one per message.** The guard *is* the
/// record; a consumer built per message remembers nothing. Replicas share a
/// store through [replayGuard].
final class TspConsumer {
  TspConsumer({
    required this.recipient,
    VidResolver? resolver,
    this.allowedSenders,
    ReplayGuard? replayGuard,
    this.replayProtection = true,
    this.freshness = FreshnessPolicy.consequential,
    this.proofVerifier,
    this.payloadValidator,
    DateTime Function() clock = DateTime.now,
  }) : resolver = resolver ?? SsiVidResolver(),
       replayGuard = replayGuard ?? InMemoryReplayGuard(),
       _clock = clock;

  /// The identity messages are addressed to, and replies are sealed from.
  final PrivateVid recipient;

  /// Resolves a sender VID to the public VID whose signature is checked. The
  /// default resolves `did:key`, `did:peer`, `did:web` and `did:webvh`.
  final VidResolver resolver;

  /// When set, only these sender VIDs are opened; checked against the message's
  /// named sender before it is resolved or opened.
  final Set<String>? allowedSenders;

  /// The §7.2 item 11 record. The in-process default is wrong for a replicated
  /// consumer: each replica would execute the same document once.
  final ReplayGuard replayGuard;

  /// Keep the record. Turn off only where every task handled declares repeated
  /// execution safe.
  final bool replayProtection;

  /// The acceptance window, and so also the record's retention.
  final FreshnessPolicy freshness;

  /// Verifies in-band proofs. Null refuses a proof-bearing document with
  /// `malformedRequest` rather than ignoring the proof. TSP authenticates the
  /// sender end-to-end, so a proof may be omitted over this binding in direct
  /// and nested modes (§5.3) — unless the specification requires one.
  final ProofVerifier? proofVerifier;

  /// Validates payloads against their schema (§7.2 item 2).
  final PayloadValidator? payloadValidator;

  final DateTime Function() _clock;

  /// Resolve the sender, open [wire], and run it through §7.2 for the
  /// specification [spec].
  ///
  /// [decode] reads the request payload and [encode] writes it back (for the
  /// replay digest). [encodeResponse] writes [handler]'s response payload. The
  /// handler returns the response document — build it with `respondWith` — or
  /// null for a specification that defines no success response, or throws a
  /// `Refusal`.
  ///
  /// Throws [TspEnvelopeException] for a message that never reached the pipeline
  /// (§4): nothing can be sent back for those.
  Future<Received<R>> receive<P, R>(
    Uint8List wire, {
    required SpecPolicy spec,
    required P Function(Map<String, dynamic> json) decode,
    required Object? Function(P payload) encode,
    required Object? Function(R payload) encodeResponse,
    required Handler<P, R> handler,
  }) async {
    final named = advertisedSender(wire);
    if (named == null) {
      throw const TspEnvelopeException(
        TspFailure.notForThisReceiver,
        'not a TSP message this consumer can read',
      );
    }
    if (allowedSenders != null && !allowedSenders!.contains(named)) {
      throw TspEnvelopeException(
        TspFailure.senderNotAllowed,
        'sender $named is not on this consumer\'s allowlist',
      );
    }
    final PublicVid sender;
    try {
      sender = await resolver.resolve(named);
    } on Object catch (e) {
      throw TspEnvelopeException(
        TspFailure.notForThisReceiver,
        'cannot resolve sender $named: $e',
      );
    }
    final unpacked = await unpackTrustTask(
      wire,
      recipient: recipient,
      sender: sender,
    );
    return consume<P, R>(
      unpacked,
      spec: spec,
      decode: decode,
      encode: encode,
      encodeResponse: encodeResponse,
      handler: handler,
    );
  }

  /// [receive] for a document already unpacked — for a caller that routes on
  /// `unpacked.document['type']` before choosing the specification.
  Future<Received<R>> consume<P, R>(
    UnpackedTrustTask unpacked, {
    required SpecPolicy spec,
    required P Function(Map<String, dynamic> json) decode,
    required Object? Function(P payload) encode,
    required Object? Function(R payload) encodeResponse,
    required Handler<P, R> handler,
  }) async {
    final transport = unpacked.transport;
    final json = unpacked.document;
    final now = _clock();

    Received<R> refused(TrustTaskDocument<Object?> doc, RejectReason reason) {
      final error = reject(transport, doc, newUrnUuid(), reason);
      return Received<R>(
        outcome: error == null ? Suppressed<R>(reason) : Rejected<R>(error),
        transport: transport,
        reply: error?.toJson((p) => p.toJson()),
      );
    }

    // unpackTrustTask guarantees string `id` and `type`, so a body whose other
    // members fail to parse can still be answered — the sender is authenticated.
    TrustTaskDocument<Object?> envelopeDoc;
    try {
      envelopeDoc = TrustTaskDocument<Object?>.fromJson(json, (p) => p);
    } on Object {
      String? str(String key) =>
          json[key] is String ? json[key] as String : null;
      envelopeDoc = TrustTaskDocument<Object?>(
        id: json['id'] as String,
        type: json['type'] as String,
        issuer: str('issuer'),
        recipient: str('recipient'),
        threadId: str('threadId'),
        parentThreadId: str('parentThreadId'),
        payload: null,
      );
      return refused(
        envelopeDoc,
        const RejectReason(
          code: StandardCode.malformedRequest,
          message: 'body is not a well-formed Trust Task document',
        ),
      );
    }

    final TrustTaskDocument<P> doc;
    try {
      doc = TrustTaskDocument<P>.fromJson(
        json,
        (p) => decode(p! as Map<String, dynamic>),
      );
    } on Object {
      return refused(
        envelopeDoc,
        const RejectReason(
          code: StandardCode.malformedRequest,
          message: 'payload does not match the specification',
        ),
      );
    }

    final verifier = proofVerifier;
    final validator = payloadValidator;
    final ConsumeOutcome<R> outcome;
    try {
      outcome = await consumeInbound<P, R>(
        transport: transport,
        spec: spec,
        proofPolicy: verifier == null
            ? const ProofPolicy.rejectIfPresent()
            : ProofPolicy.verify(_AsReceived(verifier, json)),
        payloadPolicy: validator == null
            ? const PayloadPolicy.acceptUnvalidated()
            : PayloadPolicy.validate(validator),
        checks: ConsumeChecks(
          freshness: freshness,
          replay: replayProtection
              ? ReplayPolicy.guarded(replayGuard)
              : const ReplayPolicy.notConsequential(),
        ),
        doc: doc,
        myVid: transport.local ?? '',
        now: now,
        newErrorId: newUrnUuid,
        payloadToJson: encode,
        handler: handler,
      );
    } on Object {
      // The handler threw something that is not a Refusal. Release the claim
      // before the error propagates: holding it would absorb every re-forward
      // of this document in silence until the record expires.
      if (replayProtection) {
        try {
          await replayGuard.release(doc.id, documentDigest(doc, encode));
        } on Object {
          /* best-effort */
        }
      }
      rethrow;
    }

    final reply = switch (outcome) {
      Handled(:final response) => response.toJson(encodeResponse),
      Rejected(:final error) => error.toJson((p) => p.toJson()),
      Accepted() || Suppressed() => null,
      DuplicateOutcome(:final priorResponse) => switch (priorResponse) {
        final TrustTaskDocument<R> d => d.toJson(encodeResponse),
        final Map<String, dynamic> m => m,
        _ => null,
      },
    };
    return Received<R>(outcome: outcome, transport: transport, reply: reply);
  }

  /// Seal [received]'s reply into a TSP message back to the authenticated
  /// sender (§6).
  Future<Uint8List> packReply(Received<Object?> received) async {
    final reply = received.reply;
    final peer = received.transport.peer;
    if (reply == null || peer == null) {
      throw StateError('nothing to reply with, or nobody to reply to');
    }
    final receiver = await resolver.resolve(peer);
    return packTrustTask(reply, sender: recipient, receiver: receiver);
  }
}

/// Verifies the document as it arrived rather than as the typed decode would
/// re-encode it, which drops members the generated type does not model and so
/// breaks a valid signature.
final class _AsReceived implements ProofVerifier {
  const _AsReceived(this.inner, this.json);

  final ProofVerifier inner;
  final Map<String, dynamic> json;

  @override
  Future<bool> verify(Map<String, dynamic> doc) => inner.verify(json);
}

File: trust-tasks-go/tsp/consumer.go
package tsp

import (
	"context"
	"encoding/json"
	"errors"
	"time"

	tsp "github.com/affinidi/affinidi-tsp-go"
	tt "github.com/trustoverip/dtgwg-trust-tasks-tf/trust-tasks-go/trusttasks"
)

// Consumer is the guarded inbound path for the TSP binding: open the sealed
// message, then run SPEC §7.2 with the duplicate-execution record of item 11 and
// the freshness bound it needs — both on by default.
//
// §7 records that TSP data messages "do not inherently prevent replay", and
// under §5's routed and nested carriage an intermediary may re-send the sealed
// inner message. So an ordinary re-forward — no attacker involved — executes a
// consequential task twice unless the consumer keeps the record. The record is
// keyed on the document id, never the TSP envelope, which is resealed with fresh
// material on every send (§7.1).
//
// One consumer per process, not one per message. The guard is the record; a
// consumer built per message remembers nothing. Replicas share a store through
// Guard. Build one with [NewConsumer].
type Consumer struct {
	// Recipient is the identity messages are addressed to, and replies are
	// sealed from. Required.
	Recipient *tsp.PrivateIdentity

	// Guard is the §7.2 item 11 record. The in-process default is wrong for a
	// replicated consumer: each replica would execute the same document once.
	Guard tt.ReplayGuard

	// ReplayProtection keeps the record. True by default; turn off only where
	// every task handled declares repeated execution safe.
	ReplayProtection bool

	// Freshness is the acceptance window, and so also the record's retention.
	Freshness tt.FreshnessPolicy

	// ProofVerifier verifies in-band proofs. Nil refuses a proof-bearing
	// document with malformedRequest rather than ignoring the proof. TSP
	// authenticates the sender end-to-end, so a proof may be omitted over this
	// binding in direct and nested modes (§5.3) — unless the spec requires one.
	ProofVerifier tt.ProofVerifier

	// PayloadValidator validates payloads against their schema (§7.2 item 2).
	PayloadValidator tt.PayloadValidator

	// NewID mints response and error document ids. Defaults to [NewURNUUID].
	NewID func() string

	// Now supplies the clock. Defaults to time.Now.
	Now func() time.Time
}

// NewConsumer returns a Consumer with the record on, an in-process guard and the
// consequential freshness policy — the safe defaults for a consumer that has not
// thought about item 11. Override the fields on the returned value as needed.
func NewConsumer(recipient *tsp.PrivateIdentity) *Consumer {
	return &Consumer{
		Recipient:        recipient,
		Guard:            tt.NewInMemoryReplayGuard(0),
		ReplayProtection: true,
		Freshness:        tt.ConsequentialFreshness(),
		NewID:            NewURNUUID,
		Now:              time.Now,
	}
}

func (c *Consumer) newID() string {
	if c.NewID != nil {
		return c.NewID()
	}
	return NewURNUUID()
}

func (c *Consumer) now() time.Time {
	if c.Now != nil {
		return c.Now()
	}
	return time.Now()
}

// Received is what [Receive] or [Consume] made of one TSP message.
type Received[R any] struct {
	// Outcome is the §7.2 outcome, exactly as [tt.ConsumeInbound] defines it.
	Outcome tt.ConsumeOutcome[R]
	// Transport is who sent it (authenticated) and who received it.
	Transport Transport
	// Reply is the document to seal back, as JSON, or nil when nothing is to be
	// sent: a fire-and-forget success, a suppressed identityMismatch (§8.1), or a
	// duplicate with no retained response (§7.1). Seal it with [PackReply].
	Reply json.RawMessage
}

// Receive opens wire under the resolved sender identity and runs it through
// §7.2 for the specification spec.
//
// The Go TSP library resolves no VIDs, so the caller supplies the sender
// identity — look it up from [AdvertisedSender] in your own store. decode reads
// the request payload from JSON. handler returns the response document (build it
// with [tt.RespondWith]) or nil for a fire-and-forget spec, or a [tt.Refusal]
// error.
//
// It returns an *EnvelopeError for a message that never reached the pipeline
// (§4): nothing can be sent back for those.
func Receive[P, R any](
	ctx context.Context,
	c *Consumer,
	wire []byte,
	sender *tsp.Identity,
	spec tt.SpecPolicy,
	decode func(json.RawMessage) (P, error),
	handler tt.Handler[P, R],
) (*Received[R], error) {
	unpacked, err := UnpackTrustTask(wire, c.Recipient, sender)
	if err != nil {
		return nil, err
	}
	return Consume[P, R](ctx, c, unpacked, spec, decode, handler)
}

// Consume runs an already-unpacked document through §7.2 — for a caller that
// routes on the document's type before choosing the specification.
func Consume[P, R any](
	ctx context.Context,
	c *Consumer,
	unpacked *Unpacked,
	spec tt.SpecPolicy,
	decode func(json.RawMessage) (P, error),
	handler tt.Handler[P, R],
) (*Received[R], error) {
	transport := unpacked.Transport
	now := c.now()

	refused := func(doc *tt.Document[json.RawMessage], reason tt.RejectReason) *Received[R] {
		errResp := tt.Reject(transport, doc, c.newID(), reason, nil)
		out := tt.ConsumeOutcome[R]{}
		var reply json.RawMessage
		if errResp == nil {
			out.Kind = tt.OutcomeSuppressed
			out.Reason = &reason
		} else {
			out.Kind = tt.OutcomeRejected
			out.Error = errResp
			reply, _ = json.Marshal(errResp)
		}
		return &Received[R]{Outcome: out, Transport: transport, Reply: reply}
	}

	// A raw view for the paths that answer before the typed decode. UnpackTrustTask
	// guarantees string id and type, so this always parses.
	var raw tt.Document[json.RawMessage]
	if err := json.Unmarshal(unpacked.Document, &raw); err != nil {
		return &Received[R]{
			Outcome:   tt.ConsumeOutcome[R]{Kind: tt.OutcomeRejected},
			Transport: transport,
		}, err
	}

	payload, err := decode(raw.Payload)
	if err != nil {
		return refused(&raw, tt.RejectReason{
			Code:    tt.CodeMalformedRequest,
			Message: "payload does not match the specification",
		}), nil
	}
	doc := documentWith(raw, payload)

	proofPolicy := tt.ProofPolicy{Kind: tt.ProofRejectIfPresent}
	if c.ProofVerifier != nil {
		proofPolicy = tt.ProofPolicy{
			Kind:     tt.ProofVerify,
			Verifier: asReceived{inner: c.ProofVerifier, json: unpacked.Document},
		}
	}
	payloadPolicy := tt.PayloadPolicy{Kind: tt.PayloadAcceptUnvalidated}
	if c.PayloadValidator != nil {
		payloadPolicy = tt.PayloadPolicy{Kind: tt.PayloadValidate, Validator: c.PayloadValidator}
	}
	checks := tt.ConsumeChecks{Freshness: c.Freshness, Replay: tt.ReplayPolicy{Kind: tt.ReplayNotConsequential}}
	if c.ReplayProtection {
		checks.Replay = tt.ReplayPolicy{Kind: tt.ReplayGuarded, Guard: c.Guard}
	}

	outcome, err := tt.ConsumeInbound(ctx, tt.ConsumeOptions[P, R]{
		Transport:     transport,
		Spec:          spec,
		ProofPolicy:   proofPolicy,
		PayloadPolicy: payloadPolicy,
		Checks:        checks,
		Doc:           doc,
		MyVID:         transport.Local,
		Now:           now,
		NewErrorID:    c.newID,
		Handler:       handler,
	})
	if err != nil {
		// A handler error that is not a Refusal, or a misconfiguration. Release
		// the claim so a re-forward is not absorbed in silence, then surface it.
		if c.ReplayProtection {
			if digest, derr := tt.DocumentDigest(doc); derr == nil {
				_ = c.Guard.Release(ctx, doc.ID, digest)
			}
		}
		return nil, err
	}

	var reply json.RawMessage
	switch outcome.Kind {
	case tt.OutcomeHandled:
		reply, _ = json.Marshal(outcome.Response)
	case tt.OutcomeRejected:
		reply, _ = json.Marshal(outcome.Error)
	case tt.OutcomeDuplicate:
		reply = outcome.PriorResponse
	}
	return &Received[R]{Outcome: outcome, Transport: transport, Reply: reply}, nil
}

// PackReply seals a reply document back to the authenticated sender (§6).
func PackReply(sender *tsp.PrivateIdentity, receiver *tsp.Identity, reply json.RawMessage) ([]byte, error) {
	if len(reply) == 0 {
		return nil, errors.New("tsp: nothing to reply with")
	}
	return PackTrustTask(reply, sender, receiver)
}

// documentWith rebuilds a typed document from the raw view and the decoded
// payload, preserving every other member.
func documentWith[P any](raw tt.Document[json.RawMessage], payload P) *tt.Document[P] {
	return &tt.Document[P]{
		ID:             raw.ID,
		ThreadID:       raw.ThreadID,
		ParentThreadID: raw.ParentThreadID,
		Ceremony:       raw.Ceremony,
		Type:           raw.Type,
		Issuer:         raw.Issuer,
		Recipient:      raw.Recipient,
		IssuedAt:       raw.IssuedAt,
		ExpiresAt:      raw.ExpiresAt,
		Payload:        payload,
		Context:        raw.Context,
		Proof:          raw.Proof,
		Extra:          raw.Extra,
	}
}

// asReceived verifies the document as it arrived rather than as the typed decode
// would re-encode it, which drops members the generated type does not model and
// so breaks a valid signature.
type asReceived struct {
	inner tt.ProofVerifier
	json  json.RawMessage
}

func (a asReceived) Verify(ctx context.Context, _ json.RawMessage) error {
	return a.inner.Verify(ctx, a.json)
}
